rfd/0206-stricter-linting-for-the-frontend.md

```diff
@@ -0,0 +1,411 @@
+promise in a place where it is not expected.
+
+### @typescript-eslint/no-non-null-assertion
+
+This prevents usage of `!` in the codebase to assert to the compiler that a value is not `null` or `undefined`. This can
+lead to runtime errors if the value is actually `null` or `undefined`. Instead of using `!`, the code should be
+refactored to
+ensure that the value is not `null` or `undefined` before using it.
+
+### and more...
+
+There are a lot of other rules in ESLint's TypeScript plugin that can help to improve the quality of the codebase. The
+full list [is here](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/strict-type-checked.ts).
+
+There are also [stylistic type checked rules](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/stylistic-type-checked.ts)
+which can help to improve the consistency and readability of the codebase.
+
+## Migration plan
+
+These rules cannot be enabled all at once, or even one-by-one, as there would be thousands of errors which would be hard
+to fix in one PR and backport, and could also potentially disrupt other pull requests in flight.
+
+Instead, we can split off the frontend codebase (design, shared, teleport, teleterm) into a separate directory, and
+migrate code over to the new directory, fixing the strict errors as we go. This would make backporting a lot easier,
+as the new migrated code would not clash with any existing code in the backport branches, and the code that is
+backported into the old codebase would be import path updates.
+
+Due to the way the TypeScript compiler works, it is not possible to have strictly typed code import non-strictly typed
+code, without also strictly checking the non-strictly typed code. This means that we cannot have a mix of strict and
+non-strict code in the same directory.
+
+It is possible for the non-strict code to import strictly typed code, so as we migrate parts of the codebase over, the
+existing codebase can import and use the code from the new location.
+
+Once the migration is complete, we would be left with the codebase being strictly typed.
+
+Whilst we introduce a new design component system (outlined
+in [RFD 196](https://github.com/gravitational/teleport/pull/53157)) (something else that would also require the
+separation of code into a new directory during a migration), it
+is a great opportunity to also add strict type checking.
+
+### Current directory structure
+
+```
+- e
+  - web
+    - teleport
+- web
+  - packages
+    - design
+    - shared
+    - teleport
+    - teleterm
+```
+
+### Proposed directory structure
+
+```
+- e
+  - web
+    - new
+    - teleport
+- web
+  - packages
+    - new # maybe a better name?
+      - design
+      - shared
+      - teleport
+      - teleterm
+
+    - design
+    - shared
+    - teleport
+    - teleterm
+```
+
+### Importing between codebases
```

Comment:
Do we need to account for code in teleport.e importing legacy code from old locations?

Let's take the existing [`<Switchback>` component from teleport.e](https://github.com/gravitational/teleport.e/blob/d39f9eaf9f4e54185091bfa0ec93b114b022a4e9/web/teleport/src/Banner/Switchback/Switchback.tsx) that imports `shared/utils/getDurationText`. Say I refactored `getDurationText` to use strict mode and I move it to the new location.

There are two things that both are a huge pain here:

1) Breaking the teleport.e build.
2) Having to send a PR to teleport.e to address the breaking change.

In cases where a legacy function/component is used in teleport.e, I believe we could leave a re-export at the legacy path with the `@deprecated` JSDoc tag so that teleport.e isn't broken?